SignalServiceKit/src/Messages/Attachments/TSAttachmentStream.m

```diff
@@ -224,6 +229,25 @@ - (nullable NSString *)filePath
     return [[[self class] attachmentsFolder] stringByAppendingPathComponent:self.localRelativeFilePath];
 }
 
+- (nullable NSString *)thumbnailPath
+{
+    NSString *filePath = self.filePath;
+    if (!filePath) {
+        OWSFail(@"%@ Attachment missing local file path.", self.logTag);
+        return nil;
+    }
+
+    if (!self.isImage && !self.isVideo && !self.isAnimated) {
+        return nil;
+    }
+
+    NSString *filename = filePath.lastPathComponent.stringByDeletingPathExtension;
+    NSString *containingDir = filePath.stringByDeletingLastPathComponent;
+    NSString *newFilename = [filename stringByAppendingString:@"-thumbnail"];
```

Comment:
Hmm.  A while back (Spring/Summer 2017?) we switched from storing all attachments in the same directory to storing each within its own subdirectory.  The motive was to let attachments use their "source" filename when possible without having two attachments with the same "source" filename overwrite each other.  But legacy attachments _were not_ moved into subdirectories.

So, this scheme is safe for "new-style" attachments but not for legacy attachments.  i.e. if there were existing "a.jpg" and "a-thumbnail.jpg" attachments, the thumbnail for the first might overwrite the actual second - or the second might be used as the thumbnail for the first.  

I'm not terribly concerned about this so those attachments are really old, so long as:

* We use a more obscure suffix than "-thumbnail", e.g. "-signal-ios-thumbnail".
* We ensure that in the case of a name collision we never overwrite a real attachment.

Reply:
Great point!

> We use a more obscure suffix than "-thumbnail", e.g. "-signal-ios-thumbnail".

Done!

> We ensure that in the case of a name collision we never overwrite a real attachment.

Already handled:

```
if ([[NSFileManager defaultManager] fileExistsAtPath:thumbnailPath]) {
```